### DISCUSSION TITLE
Errors building PSI file

### DISCUSSION THREAD

**Discussion by georgeinva2004** [zwave-js/ZWaveJS.NET#74] [General]
Finally getting around to building PSI files.  I'm on Windows and have NPM installed (please keep in mind, I'm not a node.js person :) )
Followed the directions in the README, and getting a number of errors.

**DeprecationWarning:  The 'punycode" module is deprecated.
Warning Cannot resolve 'packageJsonPath"
Warning Cannot resolve 'adapters[adapter'**
And a whole slew of Failed to make bytecode node2-x64 for various files that look like this:
 **Failed to make bytecode node22-x64 for file C:\snapshot\PSI\node_modules\axios\index.js**

Interestingly, there is no directory C:\snapshot\...    

**The server.exe does get created, however, it throws the following error:
node:internal/modules/cjs/loader:1254
  throw err;
  ^

Error: Cannot find module 'C:\snapshot\PSI\server.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1251:15)
    at Function._resolveFilename (pkg/prelude/bootstrap.js:1955:46)
    at Module._load (node:internal/modules/cjs/loader:1077:27)
    at Function.runMain (pkg/prelude/bootstrap.js:1983:12)
    at node:internal/main/run_main_module:30:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v22.9.0**

Any thoughts?  I'm sure it's my lack of understanding of Node, but I can't seem to find any config files that might have incorrect settings.

**Comment by marcus-j-davies** [comment]
Woah!

I tried finding out why the recent email - but then saw this (Sorry - work is killing me)

Most are warnings that can be ignored (some things don't need compiling)
The mysterious directory `snapshot` refers to the virtualised  filesystem, its part of the magic that makes it all possible - the directory won't actually exist, and you can ignore the fact it's not in your file system.

The inability to run the PSI however, might be related to a bug in the executable packager + recent versions of Node JS.

Are you on Discord?

**Comment by georgeinva2004** [comment]
I'm not, sorry.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by marcus-j-davies** [comment]
Are you using the latest V4 branch (still to be released)?

(lots have been changed  / fixed)

if so....

Install Node JS 18 : https://nodejs.org/dist/v18.20.4/node-v18.20.4-x64.msi (remove v22)
Pull down the Repo (V4 Branch)

Under a privileged command prompt (it's important to work in the `./PSI` directory in command prompt):
```
cd ./PSI
npm install && npm run buld
```

it will output to `./dist/server.exe` rename it to `server.psi`

This will need to be used with the latest library (also in the same Branch)

Later versions of Node JS (Say 22) have issues with the builder library, it produces bad binaries (on windows)
it has been fixed upstream - but I haven't had the chance yet to migrate

**Reply by georgeinva2004** [reply]
Okay.  I think I figured it out.  An executable is indeed generated.  When I run it, I get an error that wasn't very useful.  Soooo, I ran the Server.js file directly using Node and saw the error.  It was related to not finding environment variables.  Looked at the ZWaveJS.NET source code and saw that config info is passed via environment.  That's when I realized I needed to just try the generated PSI file in my application.  Everything appears to be working!!!

FYI, I am using the latest 4.0.0 branch and backed down to Node v20.  I suppose I could have just installed v18, but I was trying to see how far back I needed to go before it worked.  Good news is that v20 appears to work!  It is good to know that those warnings are not critical.  Thanks for explaining the "snapshot" directory.  That was a head scratcher...

Thanks again for all the good work!

**Comment by marcus-j-davies** [comment]
Good to know!

There is an improved build script, that I will soon merge into v4

look at the branch **fix-up-smart-smart**
It has the improved build script + some further .NET code changes  (im still playing with it)

**Reply by georgeinva2004** [reply]
I will
